### PR TITLE
Add animated gif conversion

### PR DESCRIPTION
**Pull request by hugodelahousse**
This adds support for animated gif 🎉 

If the original file & the target are gifs, each frame in the gif will be resized
I didn't manage to do it in along with the other images, but I think it's enough of a special case to warrant a separate function

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,9 +1,11 @@
 #[macro_use]
 extern crate serde_derive;
 use image::codecs::jpeg::JpegDecoder;
+use image::gif::{GifDecoder, GifEncoder};
```
Comment (cagerton) [inline src/lib.rs]: nit: `image::gif` is deprecated so this should probably be `image::codecs::gif::...`

File: src/lib.rs [hunk 2]
```diff
@@ -118,9 +122,82 @@ fn get_orientation(input: &[u8]) -> Result<u32, exif::Error> {
...
+            frame,
+            resized_w,
+            resized_h,
+            request.resize_op == ResizeType::Crop,
+            request.scale_filter,
+        )),
+        _ => frame,
+    });
+
+    let mut out: Vec<u8> = Vec::new();
```
Comment (cagerton) [inline src/lib.rs]: nit: I think this should do the same thing
```suggestion
    let mut out = vec![];
```

**Review by cagerton — COMMENTED** [review]
Thanks for the PR. :)

I took some time to test this out and found a few issues when testing with some common gifs.  

Party parrot original: party_parrot and converted: party_parrot  It looks like this doesn't preserve the background color and uses the "keep" frame disposal flags which causes previous frames to persist. Past that, the repeat extension isn't set which means the image doesn't loop.
(source: https://emoji.slack-edge.com/T03S7UXFC/fiesta-parrot/a4d99d543d7f1b47.gif)



And a test image with partial frames. Original: animated_transparent_frame_restoreprev_loop. Converted: animated_transparent_frame_restoreprev_loop. Aside from the lack of repeat, this one looks generally okay.  Unfortunately the image size increases markedly, from 685 bytes to around 5kb.  Digging in a bit I think a significant contributor here is the fact that frames are expanded to full size even if they're originally only cover part of the image.  If I'm reading the code correctly, I think that's due to a design decision for the high level image crate's gif codec api https://github.com/image-rs/image/blob/787b655d737594a8331d06d099cb59c0305aa997/src/codecs/gif.rs#L113-L117
(image source: https://peterdn.com/post/2020/10/25/using-imagemagick-to-create-a-gif-test-suite/)

FWIW, I did spend some time getting a prototype working using the lower-level gif crate directly so I think this is still doable.  I'll follow up on that when time permits.

**Comment by hugodelahousse** [comment]
@cagerton Could you share your prototype with the lower-level API ? I can contribute if you don't have time to work on it
GraphicsMagick has trouble resizing gifs with partial frames in production, so it'd be good to have a fix directly from here
Original:
no_baymax
Server resize:
no_baymax_bad

**Comment by cagerton** [comment]
> @cagerton Could you share your prototype with the lower-level API ? I can contribute if you don't have time to work on it
> GraphicsMagick has trouble resizing gifs with partial frames in production, so it'd be good to have a fix directly from here
> Original:
> no_baymax
> Server resize:
> no_baymax_bad

beta

**Comment by cagerton** [comment]
> > @cagerton Could you share your prototype with the lower-level API ? I can contribute if you don't have time to work on it
> > GraphicsMagick has trouble resizing gifs with partial frames in production, so it'd be good to have a fix directly from here
> > Original:
> > no_baymax
> > Server resize:
> > no_baymax_bad
> 
> beta

I was a bit surprised that imagemagick&graphicsmagick dropped the ball here, tbh

**Comment by cagerton** [comment]
Closing this since gif support has been not been merged.

**Pull request closed without merging**